### PR TITLE
feat: Implement ProtocolSim for VMPoolstate get_spot_prices

### PR DESCRIPTION
**Pull request by dianacarvalho1**
Add spot_prices as an attribute and rename get_spot_prices (from VMPoolstate) to set_spot_prices. This way set_spot_prices needs to be called at some point by the user every time that the spot prices change. The get_spot_prices from ProtocolSim will only read from this cached spot prices

### CODE REVIEW

File: src/protocol/vm/state.rs [hunk 1]
```diff
@@ -535,6 +543,62 @@ impl VMPoolState<PreCachedDB> {
...
+impl ProtocolSim for VMPoolState<PreCachedDB> {
+    fn fee(&self) -> f64 {
+        todo!()
+    }
+
+    fn spot_price(&self, base: &ERC20Token, quote: &ERC20Token) -> f64 {
+        *self
+            .spot_prices
+            .get(&(base.clone(), quote.clone()))
+            .expect("Spot price not found")
```
[the remark this answers is not in the document]
Comment (tamaralipows) [reply src/protocol/vm/state.rs:555]: Yes there are a few instances if something is really wrong but we pass the error to users anyway, so I think we should unify about what we are going to do here. Please see this comment:

https://github.com/propeller-heads/tycho-simulation/pull/12#discussion_r1808498053

File: src/protocol/vm/state.rs [hunk 2]
```diff
@@ -535,6 +543,62 @@ impl VMPoolState<PreCachedDB> {
...
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn eq(&self, other: &dyn ProtocolSim) -> bool {
+        if let Some(other_state) = other
+            .as_any()
+            .downcast_ref::<VMPoolState<PreCachedDB>>()
+        {
+            self.id == other_state.id
```
[the remark this answers is not in the document]
Comment (tamaralipows) [reply src/protocol/vm/state.rs:595]: then I think it should be fine

File: src/protocol/vm/state.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (louise-poole) [reply src/protocol/vm/state.rs:555]: We should defs return an error. A user might decide if spot prices fail they just remove that pool and continue with all other pools that are working. We shouldn't crash their whole program for this. Better to let the user decide is the error is worth crashing for or not.

[the remark this answers is not in the document]
Comment (dianacarvalho1) [reply src/protocol/vm/state.rs:555]: ok but then we need to change the interface of the ProtocolSim for spot prices. but yes I understand the reasoning. I will change it!

**Pull request merged**